Objects/typeobject.c

```diff
@@ -11319,6 +11334,30 @@ add_operators(PyTypeObject *type)
 }
 
 
+int
+PyType_Freeze(PyTypeObject *type)
+{
+    // gh-121654: Check the __mro__ instead of __bases__
+    PyObject *mro = type_get_mro(type, NULL);
+    if (!PyTuple_Check(mro)) {
+        Py_DECREF(mro);
+        PyErr_SetString(PyExc_TypeError, "unable to get the type MRO");
+        return -1;
+    }
```

Comment:
How difficult to add test for this case?

Reply:
I'm not sure that it's possible to create a type with a MRO which is not a tuple. Using a metaclass, it's possible to override the MRO, but internally, Python converts `mro()` result into a tuple.

I added this check since `type_get_mro()` returns `None` if `type->tp_mro` is `NULL`. But I have no idea how to create a type with `type->tp_mro=NULL`. It's more a sanity check.